spring-kafka/src/main/java/org/springframework/kafka/listener/TimestampedException.java

```diff
@@ -34,14 +34,36 @@
 
 	private final long timestamp;
 
+	@Deprecated
 	public TimestampedException(Exception ex, Clock clock) {
 		super(ex.getMessage(), ex);
 		this.timestamp = Instant.now(clock).toEpochMilli();
 	}
 
+	@Deprecated
 	public TimestampedException(Exception ex) {
 		super(ex.getMessage(), ex);
-		this.timestamp = Instant.now(Clock.systemDefaultZone()).toEpochMilli();
+		this.timestamp = Instant.now().toEpochMilli();
+	}
+
+	/**
+	 * Creates an instance with the timestamp of when it was thrown and its cause.
+	 * @param ex the exception cause.
+	 * @param timestamp the timestamp of when the exception was thrown.
+	 */
+	public TimestampedException(Exception ex, long timestamp) {
+		super("Exception thrown at " + Instant.ofEpochMilli(timestamp), ex);
+		this.timestamp = timestamp;
+	}
+
+	/**
+	 * Creates an instance with the timestamp of when it was thrown and its cause.
+	 * @param ex the exception cause.
+	 * @param now the Instant of when the exception was thrown.
+	 */
+	public TimestampedException(Exception ex, Instant now) {
```

Comment:
Why just plain `long timestamp` not enough? 
Why end-user cannot do their `Instant.toEpochMilli()` when they use our `TimestampedException(Exception ex, long timestamp)` ?

Reply:
TBH it's been some time since I've seen a `Date` for anything in Java code 😄.

Those new APIs are a real improvement over the older ones, [this article](https://www.baeldung.com/java-8-date-time-intro) does a good enough overview on them. The first ones are intensely used in production code these days.

I'd keep the `long` based one because there are places where you get a `long from epoch` timestamp, for example Kafka headers. So if users already have a `long` they could just use it. But of course, this is all theoretical - I'm ok with keeping only either of them.